### PR TITLE
refactor: restructure packages to improve extensibility

### PR DESCRIPTION
**Pull request by fgreinacher**
This change splits the System.IO.Abstractions package into TestableIO.System.IO.Abstractions and TestableIO.System.IO.Abstractions.Wrappers. The former contains just the interfaces and the latter the default wrapper implementations. The package System.IO.Abstractions still exists as meta package to reduce impact on existing users. The same applies to TestableIO.System.IO.Abstractions.TestingHelpers.

BREAKING CHANGE: This refactoring moves all types to the TestableIO.System.IO.Abstractions library. Users might need to recompile their libraries.

### CODE REVIEW

**Comment by fgreinacher** [comment]
@vbreuss Could you have a quick look at this?

File: README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ![System.IO.Abstractions](https://socialify.git.ci/TestableIO/System.IO.Abstractions/image?description=1&font=Source%20Code%20Pro&forks=1&issues=1&pattern=Charlie%20Brown&pulls=1&stargazers=1&theme=Dark)
-[![NuGet](https://img.shields.io/nuget/v/System.IO.Abstractions.svg)](https://www.nuget.org/packages/System.IO.Abstractions)
+[![NuGet](https://img.shields.io/nuget/v/TestableIO.System.IO.Abstractions.svg)](https://www.nuget.org/packages/TestableIO.System.IO.Abstractions)
 ![Continuous Integration](https://github.com/TestableIO/System.IO.Abstractions/workflows/Continuous%20Integration/badge.svg)
 [![Codacy Badge](https://app.codacy.com/project/badge/Grade/54479b054d194adfb4ff476ef0182fe0)](https://www.codacy.com/gh/TestableIO/System.IO.Abstractions/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=TestableIO/System.IO.Abstractions&amp;utm_campaign=Badge_Grade)
```
Comment (vbreuss) [inline README.md:4]: The codacy badge seems incorrect on the `refactor/packaging` branch:
image

File: tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj [hunk 2]
```diff
@@ -24,8 +24,8 @@
     <EmbeddedResource Include="TestFiles\TestFile.txt" />
   </ItemGroup>
   <ItemGroup>
-    <ProjectReference Include="../../src/System.IO.Abstractions/System.IO.Abstractions.csproj" />
-    <ProjectReference Include="../../src/System.IO.Abstractions.TestingHelpers/System.IO.Abstractions.TestingHelpers.csproj" />
+    <ProjectReference Include="../../src/TestableIO.System.IO.Abstractions/STestableIO.ystem.IO.Abstractions.csproj" />
```
Comment (vbreuss) [inline tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj:27]:
```suggestion
    <ProjectReference Include="../../src/TestableIO.System.IO.Abstractions/TestableIO.System.IO.Abstractions.csproj" />
```

Typo in the project reference!

File: System.IO.Abstractions.sln [hunk 3]
```diff
@@ -30,16 +28,18 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "images", "images", "{F78189
 		images\icon_256x256.png = images\icon_256x256.png
 	EndProjectSection
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{C078E0B6-9747-475F-A999-B9E775DF6643}"
```
Comment (vbreuss) [inline System.IO.Abstractions.sln:31]: The "Solution Explorer" is now quite inconsistent (only one project in directory `src`):
image

I would suggest keep the functional projects on the outermost level and move the meta projects to a dedicated directory, e.g.:
image

**Review by vbreuss — CHANGES_REQUESTED** [review]

**Comment by vbreuss** [comment]
> @vbreuss Could you have a quick look at this?

@fgreinacher : I finished my review :-)

File: README.md (hunk 1 above)
Comment (fgreinacher) [reply README.md:4]: Seems this was a temporary glitch, it works now:

image

File: tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj (hunk 2 above)
Comment (fgreinacher) [reply tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj:27]: Ah, great catch. Weird that CI did not complain about this 🤔

File: tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj [hunk 4]
```diff
@@ -24,8 +24,7 @@
     <EmbeddedResource Include="TestFiles\TestFile.txt" />
   </ItemGroup>
   <ItemGroup>
-    <ProjectReference Include="../../src/System.IO.Abstractions/System.IO.Abstractions.csproj" />
-    <ProjectReference Include="../../src/System.IO.Abstractions.TestingHelpers/System.IO.Abstractions.TestingHelpers.csproj" />
+    <ProjectReference Include="../../src/TestableIO.System.IO.Abstractions.TestingHelpers/TestableIO.System.IO.Abstractions.TestingHelpers.csproj" />
```
Comment (fgreinacher) [inline tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj:27]: `TestableIO.System.IO.Abstractions` is pulled in transitively, so I removed the direct reference.

File: README.md (hunk 1 above)
Comment (vbreuss) [reply README.md:4]: Yes, you are right.
I just saw it when reading the updated README.md...

File: tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj (hunk 4 above)
Comment (vbreuss) [reply tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests/TestableIO.System.IO.Abstractions.TestingHelpers.Tests.csproj:27]: I was wondering, why the build system succeeded, despite this typo. This is even better!

File: System.IO.Abstractions.sln (hunk 3 above)
Comment (fgreinacher) [reply System.IO.Abstractions.sln]: Oh, that's indeed messy. I cleaned it up a bit but I'm only on VS Code, please have a look :)

**Comment by fgreinacher** [comment]
Thanks for your comments @vbreuss, ready for round 2️⃣

File: System.IO.Abstractions.sln (hunk 3 above)
Comment (vbreuss) [reply System.IO.Abstractions.sln]: Looks much better now:
image

Comment (vbreuss) [reply System.IO.Abstractions.sln]: What is the Benchmarks project? Should this also have the "TestableIO" prefix?

**Review by vbreuss — APPROVED** [review]
@fgreinacher : Looks good!

What is the Benchmarks project?
Should it also have the "TestableIO" prefix?

*It seems I can't resolve conversations or did I just not find the corresponding button?*

**Comment by fgreinacher** [comment]
Thanks for the quick review @vbreuss!

> What is the Benchmarks project?
> Should it also have the "TestableIO" prefix?

It contains some performance benchmarks. I will also add the prefix there!

> It seems I can't resolve conversations or did I just not find the corresponding button?

I guess that's because you're not member of the organization. If you're interested helping out here, just say so and I'll add you as collaborator.

**Comment by vbreuss** [comment]
> I guess that's because you're not member of the organization. If you're interested helping out here, just say so and I'll add you as collaborator.

@fgreinacher : I would be happy to help out!

**Comment by fgreinacher** [comment]
> > I guess that's because you're not member of the organization. If you're interested helping out here, just say so and I'll add you as collaborator.
> 
> @fgreinacher : I would be happy to help out!

Cool thanks, I added you to the project!

**Pull request merged**